### ISSUE TITLE
Pyright infers wrong signature for generated dataclass __init__() with mangled fields

### ISSUE THREAD

**Issue by yaxum62** [microsoft/pyright#4584]
Pyright seems does not handle dataclass with mangled fields.

Given a dataclass with private attribute:
```python
@dataclass
class Foo:
  x: int
  __y: int
```

Creating a `Foo` instance like `Foo(x=1, _Foo__y=2)`, which is accepted by python interpreter, is marked error by pyright with 2 message:
- Argument missing for parameter "__y"
- No parameter named "_Foo__y"

The suggested call `Foo(x=1, __y=2)` passed Pyright check but will actually cause TypeError in python interpreter.

Things make less sense when a dataclass only contains one private field, e.g.:
```python
@dataclass
class Foo:
  __y: int
```

Both call `Foo(__y=1)` and `Foo(_Foo__y=1)` are marked as error with message "Expected 1 more positional argument"

Pyright version: vscode python extension v2023.2.0
Python version: 3.10.9

**Comment by erictraut** [comment]
You shouldn't use private attributes with dataclasses. This is a misuse of a private attribute, which is intended to be used internally to the class, not exposed externally to the class.

I doubt if the creators of dataclasses thought about this edge case. PEP 557 doesn't mention anything about it.

Not surprisingly, mypy does not generate any errors in your first example either.

I've updated pyright to flag the use of a private attribute name as an error.

**Comment by erictraut** [comment]
This is included in pyright 1.1.293, which I just published. It will also be included in a future release of pylance.
